src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/PosixSignalRegistration.Unix.cs

```diff
@@ -0,0 +1,173 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Threading;
+
+namespace System.Runtime.InteropServices
+{
+    public sealed class PosixSignalRegistration : IDisposable
+    {
+        private readonly Action<PosixSignalContext> _handler;
+        private readonly PosixSignal _signal;
+        private bool _registered;
+        private readonly object _gate = new object();
+
+        private PosixSignalRegistration(PosixSignal signal, Action<PosixSignalContext> handler)
+        {
+            _signal = signal;
+            _handler = handler;
+        }
+
+        public static PosixSignalRegistration Create(PosixSignal signal, Action<PosixSignalContext> handler)
+        {
+            if (handler == null)
+            {
+                throw new ArgumentNullException(nameof(handler));
+            }
+            if (signal > PosixSignal.SIGHUP || signal < PosixSignal.SIGCHLD)
+            {
+                throw new IndexOutOfRangeException();
+            }
+            PosixSignalRegistration registration = new PosixSignalRegistration(signal, handler);
+            registration.RegisterFor(signal);
+            return registration;
+        }
+
+        private unsafe void RegisterFor(PosixSignal signal)
+        {
+            if (!s_initialized)
+            {
+                if (!Interop.Sys.InitializeTerminalAndSignalHandling())
+                {
+                    throw new Win32Exception();
+                }
+                s_initialized = true;
+            }
+            lock (s_registrations)
+            {
+                if (!s_registrations.TryGetValue(signal, out List<WeakReference<PosixSignalRegistration>>? signalRegistrations))
+                {
+                    if (!Interop.Sys.RegisterForPosixSignal(signal, &OnPosixSignal))
```

Comment:
I don't see the code ever removing a list from s_registrations.  Seems then when the first handler is added we'll RegisterForPosixSignal, then if it's removed the count will hit 0 and we'll UnregisterForPosixSignal, but then if another is added back in we'll never re-RegisterForPosixSignal.,

Reply:
I've added it.